keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+
+While it is possible to inline a blocklist directly into a CEL expression as a
+data literal (`!(object.metadata.name in ['blocked1', 'blocked2'])`), this quickly
+becomes problematic:
+
+- Long blocklists become unwieldy quickly in CEL expressions
+- A blocklist per scope (e.g. namespace) is inconvenient express and maintain
+
+The need to configure admission rules is common enough (see below use cases)
+that we propose configuration be a 1st class concept in the API.
+
+Since all the policy frameworks we have surveyed have configurability as a 1st
+class concept, omitting it would result in either the policy frameworks not
+adopting this enhancement (and sticking with webhooks) or somehow bypassing the
+limitation. One possible approach would be to generate a CEL expression with the
+configuration data embedded as a data literal, but we would strongly prefer not
+to have policy frameworks generating a CEL expression for each possible
+configuration.
+
+### Migration
+
+With webhooks already in large scale use in the Kubernetes ecosystem, we intend
+to prioritize capabilities that ease migration. As a concrete example, when
+migrating, having fine grained control of what validation messages are returned
+and how they are formatted can make a migration far more seamless.
+
+### Compliance
+
+In-process admission control has fundamental advantages over webhooks: it is far
+safer to use in a "fail closed" mode. With webhooks, using "fail closed" can
+negatively impact cluster availability. But "fail closed" is very valuable when
+enforcing compliance (and security). We intend to prioritize capabilities that
+make "fail closed" a safe mode of operation. As a concrete example, only
+allowing CEL expressions that pass compilation and type checking significantly
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare "zero trust" policies is
+important to compliance. By "zero trust", we mean policy rules that apply the
+principle of least privilege to newly created resources (e.g. namespaces) where
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to webhook Rules, NamespaceSelector,
+  ObjectSelector)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Automatic version conversion support (similar to webhooks MatchPolicy=Equivalent), with ability to introspect GVK of admitted resource in CEL expression.
```

Comment:
I'm in favor of providing this, but will there be some kind of feedback in the resource that indicates this is happening?  Having conversion means that webhook calls enter into the flow.

Reply:
Another thing to think about is portability -- conversion webhooks aren't always available shift left/right.

It might be worth supporting both conversion and the ability to write policies against multiple versions of the same object.

Reply:
Added to KEP.